pkg/proto/eth_transaction.go

```diff
@@ -112,42 +118,57 @@ func (tx *EthereumTransferWavesTxKind) String() string {
 type EthereumTransferAssetsErc20TxKind struct {
 	decodedData ethabi.DecodedCallData
 	Arguments   ethabi.ERC20TransferArguments
-	Asset       OptionalAsset
+	OptAsset    *OptionalAsset
 }
 
-func NewEthereumTransferAssetsErc20TxKind(decodedData ethabi.DecodedCallData, asset OptionalAsset, arguments ethabi.ERC20TransferArguments) *EthereumTransferAssetsErc20TxKind {
-	return &EthereumTransferAssetsErc20TxKind{Asset: asset, decodedData: decodedData, Arguments: arguments}
+func NewEthereumTransferAssetsErc20TxKind(decodedData ethabi.DecodedCallData, asset *OptionalAsset, arguments ethabi.ERC20TransferArguments) *EthereumTransferAssetsErc20TxKind {
+	return &EthereumTransferAssetsErc20TxKind{OptAsset: asset, decodedData: decodedData, Arguments: arguments}
 }
 
-func (tx *EthereumTransferAssetsErc20TxKind) DecodedData() *ethabi.DecodedCallData {
-	return &tx.decodedData
+func (tx *EthereumTransferAssetsErc20TxKind) DecodedData() (*ethabi.DecodedCallData, error) {
+	return &tx.decodedData, nil
+}
+func (tx *EthereumTransferAssetsErc20TxKind) Asset() (*OptionalAsset, error) {
+	if tx.OptAsset == nil {
+		return nil, errors.New("asset field of ethereum transfer assets tx is empty")
+	}
+	return tx.OptAsset, nil
 }
 
 func (tx *EthereumTransferAssetsErc20TxKind) String() string {
 	return "EthereumTransferAssetsErc20TxKind"
 }
 
 type EthereumInvokeScriptTxKind struct {
-	decodedData ethabi.DecodedCallData
+	DecodedCallData *ethabi.DecodedCallData
 }
 
-func NewEthereumInvokeScriptTxKind(decodedData ethabi.DecodedCallData) *EthereumInvokeScriptTxKind {
-	return &EthereumInvokeScriptTxKind{decodedData: decodedData}
+func NewEthereumInvokeScriptTxKind(decodedData *ethabi.DecodedCallData) *EthereumInvokeScriptTxKind {
+	return &EthereumInvokeScriptTxKind{DecodedCallData: decodedData}
 }
 
-func (tx *EthereumInvokeScriptTxKind) DecodedData() *ethabi.DecodedCallData {
-	return &tx.decodedData
+func (tx *EthereumInvokeScriptTxKind) DecodedData() (*ethabi.DecodedCallData, error) {
+	if tx.DecodedCallData == nil {
+		return nil, errors.New("ethereum invoke script tx has empty decoded data")
+	}
+	return tx.DecodedCallData, nil
 }
 
 func (tx *EthereumInvokeScriptTxKind) String() string {
 	return "EthereumInvokeScriptTxKind"
 }
 
 type EthereumTransaction struct {
-	inner           EthereumTxData
-	innerBinarySize int
-	senderPK        atomic.Value // *EthereumPublicKey
+	// Ethereum representation
+	senderPK atomic.Value   // *EthereumPublicKey
+	inner    EthereumTxData //
+
+	// Waves representation
 	TxKind          EthereumTransactionKind
```

Comment:
Use different naming for `Ethereum representation` fields and `Waves representation` fields. I think it would be better if you create special struct for valid waves representation.